Source/GSAutoLayoutEngine.m

```diff
@@ -1,60 +1,460 @@
    version 2.1 of the License, or (at your option) any later version.
-   
+
    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.
-   
+
    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free
    Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
    Boston, MA 02110 USA.
 */
 
 #import "GSAutoLayoutEngine.h"
-#include "AppKit/NSLayoutConstraint.h"
+#import "AppKit/NSLayoutConstraint.h"
+#import "GSCSConstraint.h"
+#import "GSFastEnumeration.h"
+#import "NSViewPrivate.h"
+
+@interface
+GSAutoLayoutEngine (PrivateMethods)
+
+- (GSCSConstraint *) solverConstraintForConstraint:
+    (NSLayoutConstraint *)constraint;
+
+- (void) mapLayoutConstraint: (NSLayoutConstraint *)layoutConstraint
+          toSolverConstraint: (GSCSConstraint *)solverConstraint;
+
+- (void) updateAlignmentRectsForTrackedViews;
+
+- (void) updateAlignmentRectsForTrackedViewsForSolution:
+    (GSCSSolution *)solution;
+
+- (BOOL) hasConstraintsForView: (NSView *)view;
+
+- (GSCSConstraint *) getExistingConstraintForAutolayoutConstraint:
+    (NSLayoutConstraint *)constraint;
+
+- (void) addConstraintAgainstViewConstraintsArray:
+    (NSLayoutConstraint *)constraint;
+
+- (void) removeConstraintAgainstViewConstraintsArray:
+    (NSLayoutConstraint *)constraint;
+
+- (void) addSupportingInternalConstraintsToView: (NSView *)view
+                                   forAttribute: (NSLayoutAttribute)attribute
+                                     constraint: (GSCSConstraint *)constraint;
+
+- (void) removeInternalConstraintsForView: (NSView *)view;
+
+- (void) addSolverConstraint: (GSCSConstraint *)constraint;
+
+- (void) removeSolverConstraint: (GSCSConstraint *)constraint;
+
+- (NSArray *) constraintsForView: (NSView *)view;
+
+- (NSNumber *) indexForView: (NSView *)view;
+
+- (void) notifyViewsOfAlignmentRectChange: (NSArray *)viewsWithChanges;
+
+- (BOOL) solverCanSolveAlignmentRectForView: (NSView *)view
+                                   solution: (GSCSSolution *)solution;
+
+- (void) recordAlignmentRect: (NSRect)alignmentRect
+                forViewIndex: (NSNumber *)viewIndex;
+
+- (NSRect) solverAlignmentRectForView: (NSView *)view
+                             solution: (GSCSSolution *)solution;
+
+- (NSRect) currentAlignmentRectForViewAtIndex: (NSNumber *)viewIndex;
+
+- (BOOL) updateViewAligmentRect: (GSCSSolution *)solution view: (NSView *)view;
+
+@end
 
 @implementation GSAutoLayoutEngine
 
-- (void) addConstraint: (NSLayoutConstraint *)constraint
+- (instancetype) initWithSolver: (GSCassowarySolver *)cassowarySolver;
 {
-  // FIXME Add constraint to solver
+  self = [super init];
+  if (self != nil)
+    {
+      ASSIGN(_solver, cassowarySolver);
+      RETAIN(_solver);
```

Comment:
The idea of using ASSIGN is that you won't need the RETAIN after it. The macro already takes case of the needed retains and releases. Also this change should be applied to all the places where you actually assign an object to an instance variable.

Reply:
Fixed, thanks for explanation. I take it that if releases are handled with the ASSIGN macro the dealloc methods are no longer required.

Reply:
No, you still need the RELEASE in the dealloc method. What I was referring to is that the ASSIGN macro will release any previous value in the first parameter (in your case the instance variable)

Reply:
My mistake, added them back